src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/OrdinalStringFrozenDictionary_LeftJustifiedSingleCharCaseInsensitive.cs

```diff
@@ -0,0 +1,29 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Collections.Frozen
+{
+    internal sealed class OrdinalStringFrozenDictionary_LeftJustifiedSingleCharCaseInsensitive<TValue> : OrdinalStringFrozenDictionary<TValue>
+    {
+        internal OrdinalStringFrozenDictionary_LeftJustifiedSingleCharCaseInsensitive(
+            string[] keys,
+            TValue[] values,
+            IEqualityComparer<string> comparer,
+            int minimumLength,
+            int maximumLengthDiff,
+            int hashIndex)
+            : base(keys, values, comparer, minimumLength, maximumLengthDiff, hashIndex, 1)
+        {
+        }
+
+        // This override is necessary to force the jit to emit the code in such a way that it
+        // avoids virtual dispatch overhead when calling the Equals/GetHashCode methods. Don't
+        // remove this, or you'll tank performance.
+        private protected override ref readonly TValue GetValueRefOrNullRefCore(string key) => ref base.GetValueRefOrNullRefCore(key);
+
+        private protected override bool Equals(string? x, string? y) => StringComparer.OrdinalIgnoreCase.Equals(x, y);
```

Comment:
Could we perhaps avoid creating new types using `string.Equals(x, y, StringComparison.OrdinalIgnoreCase);` here instead? We could just parameterize the frozen dictionary on a `StringComparison` input.

Note I haven't measured performance for this, but it certainly wouldn't involve any virtual calls.

Reply:
Thanks for suggesting that. I previously used `Comparer.Equals` and that was slow. I've created a PR to use `string.Equals(x, y, StringComparison.OrdinalIgnoreCase)` - https://github.com/andrewjsaid/dotnet-runtime/pull/1/files

Unfortunately this approach has a significant performance penalty

Benchmarks below
## Perf_DefaultFrozenDictionary
```
// * Summary *

BenchmarkDotNet v0.13.7-nightly.20230717.35, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 8.0.100-rc.2.23502.2
  [Host]     : .NET 8.0.0 (8.0.23.47906), X64 RyuJIT AVX2
  Job-RLAWOF : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-MDFFXR : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-APYSFT : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:EnableUnsafeBinaryFormatterSerialization=true  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1  

|                             Method |        Job |                         ...  Toolchain | Count |          Mean |        Error |       StdDev |        Median |           Min |           Max | Ratio | RatioSD | Allocated | Alloc Ratio |
|----------------------------------- |----------- |-------------------------...----------- |------ |--------------:|-------------:|-------------:|--------------:|--------------:|--------------:|------:|--------:|----------:|------------:|
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |      25.06 ns |     0.181 ns |     0.161 ns |      24.99 ns |      24.89 ns |      25.41 ns |  1.02 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |      38.10 ns |     0.260 ns |     0.231 ns |      38.18 ns |      37.50 ns |      38.30 ns |  1.56 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |      24.46 ns |     0.296 ns |     0.277 ns |      24.43 ns |      24.05 ns |      25.05 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |               |              |              |               |               |               |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |      23.55 ns |     1.057 ns |     1.218 ns |      23.31 ns |      21.93 ns |      26.13 ns |  1.09 |    0.06 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |      25.16 ns |     0.141 ns |     0.131 ns |      25.20 ns |      24.97 ns |      25.39 ns |  1.16 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |      21.73 ns |     0.214 ns |     0.200 ns |      21.79 ns |      21.33 ns |      21.93 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |               |              |              |               |               |               |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |     754.69 ns |     5.098 ns |     4.768 ns |     753.37 ns |     747.62 ns |     764.75 ns |  0.98 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |     769.36 ns |     9.790 ns |     9.157 ns |     765.15 ns |     758.89 ns |     787.10 ns |  1.00 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |     772.13 ns |     8.615 ns |     8.059 ns |     775.36 ns |     758.14 ns |     783.36 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |               |              |              |               |               |               |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |     665.25 ns |     6.411 ns |     5.997 ns |     666.57 ns |     654.81 ns |     673.53 ns |  1.00 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |     677.18 ns |     3.126 ns |     2.924 ns |     678.11 ns |     670.69 ns |     681.12 ns |  1.02 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |     662.41 ns |     3.441 ns |     3.050 ns |     661.41 ns |     659.37 ns |     670.57 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |               |              |              |               |               |               |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |   7,936.93 ns |    59.358 ns |    55.524 ns |   7,915.32 ns |   7,869.44 ns |   8,057.95 ns |  0.98 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |   8,184.02 ns |    88.753 ns |    83.019 ns |   8,188.93 ns |   7,987.79 ns |   8,315.43 ns |  1.01 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |   8,112.07 ns |    25.424 ns |    22.538 ns |   8,117.14 ns |   8,068.10 ns |   8,155.61 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |               |              |              |               |               |               |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |   7,520.96 ns |    59.546 ns |    52.786 ns |   7,517.38 ns |   7,434.73 ns |   7,614.71 ns |  1.00 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |   7,617.98 ns |    51.175 ns |    45.365 ns |   7,617.81 ns |   7,552.51 ns |   7,701.13 ns |  1.02 |    0.02 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |   7,491.18 ns |    80.914 ns |    75.687 ns |   7,491.23 ns |   7,347.80 ns |   7,609.99 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |               |              |              |               |               |               |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 | 165,876.99 ns | 1,167.865 ns | 1,035.282 ns | 165,445.61 ns | 164,378.86 ns | 167,575.40 ns |  1.01 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 | 161,778.83 ns | 1,946.199 ns | 1,725.254 ns | 161,599.36 ns | 158,910.84 ns | 165,543.24 ns |  0.99 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 | 163,557.16 ns |   521.449 ns |   462.251 ns | 163,573.44 ns | 162,717.84 ns | 164,515.23 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |               |              |              |               |               |               |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 | 131,257.56 ns | 1,171.590 ns | 1,038.584 ns | 130,709.90 ns | 130,379.64 ns | 133,406.98 ns |  0.94 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 | 131,434.81 ns | 1,010.655 ns |   895.920 ns | 131,106.86 ns | 129,988.93 ns | 133,266.95 ns |  0.94 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 | 140,172.95 ns | 1,495.260 ns | 1,325.509 ns | 140,413.91 ns | 137,378.43 ns | 142,223.73 ns |  1.00 |    0.00 |         - |          NA |
```

## Perf_LengthBucketsFrozenDictionary
```
// * Summary *

BenchmarkDotNet v0.13.7-nightly.20230717.35, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 8.0.100-rc.2.23502.2
  [Host]     : .NET 8.0.0 (8.0.23.47906), X64 RyuJIT AVX2
  Job-RLAWOF : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-MDFFXR : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-APYSFT : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:EnableUnsafeBinaryFormatterSerialization=true  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1  

|                             Method |        Job |                         ...  Toolchain | Count | ItemsPerBucket |           Mean |         Error |      StdDev |         Median |            Min |            Max | Ratio | RatioSD | Allocated | Alloc Ratio |
|----------------------------------- |----------- |-------------------------...----------- |------ |--------------- |---------------:|--------------:|------------:|---------------:|---------------:|---------------:|------:|--------:|----------:|------------:|
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |              1 |      17.426 ns |     0.2208 ns |   0.1957 ns |      17.464 ns |      17.065 ns |      17.711 ns |  1.02 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |              1 |      16.857 ns |     0.1828 ns |   0.1710 ns |      16.891 ns |      16.556 ns |      17.125 ns |  0.99 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |              1 |      17.012 ns |     0.2951 ns |   0.2761 ns |      16.997 ns |      16.583 ns |      17.504 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |              1 |       8.668 ns |     0.0762 ns |   0.0676 ns |       8.638 ns |       8.595 ns |       8.782 ns |  1.00 |    0.02 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |              1 |       8.480 ns |     0.1357 ns |   0.1269 ns |       8.489 ns |       8.239 ns |       8.682 ns |  0.98 |    0.02 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |              1 |       8.638 ns |     0.1191 ns |   0.1114 ns |       8.614 ns |       8.437 ns |       8.873 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |              5 |      59.099 ns |     0.5217 ns |   0.4880 ns |      58.914 ns |      58.637 ns |      60.013 ns |  1.06 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |              5 |      56.704 ns |     0.1669 ns |   0.1479 ns |      56.688 ns |      56.455 ns |      56.964 ns |  1.02 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |              5 |      55.501 ns |     0.3720 ns |   0.3479 ns |      55.515 ns |      55.062 ns |      56.220 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |              5 |       8.564 ns |     0.1201 ns |   0.1064 ns |       8.574 ns |       8.273 ns |       8.726 ns |  1.00 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |              5 |       8.355 ns |     0.1387 ns |   0.1298 ns |       8.382 ns |       8.075 ns |       8.571 ns |  0.98 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |              5 |       8.543 ns |     0.0336 ns |   0.0314 ns |       8.543 ns |       8.484 ns |       8.599 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |              1 |     156.324 ns |     0.5829 ns |   0.5452 ns |     156.319 ns |     154.845 ns |     157.234 ns |  1.00 |    0.00 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |              1 |     156.123 ns |     1.5892 ns |   1.4865 ns |     155.303 ns |     154.490 ns |     158.933 ns |  1.00 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |              1 |     156.253 ns |     0.4476 ns |   0.3494 ns |     156.336 ns |     155.596 ns |     156.644 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |              1 |      78.237 ns |     0.4816 ns |   0.4269 ns |      78.151 ns |      77.646 ns |      79.137 ns |  0.99 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |              1 |      76.309 ns |     0.6654 ns |   0.6224 ns |      76.586 ns |      75.189 ns |      77.293 ns |  0.96 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |              1 |      79.365 ns |     0.8172 ns |   0.7244 ns |      79.414 ns |      78.223 ns |      80.247 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |              5 |     705.449 ns |    13.7909 ns |  15.3285 ns |     702.310 ns |     688.315 ns |     739.583 ns |  1.04 |    0.04 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |              5 |     685.888 ns |    14.7915 ns |  17.0339 ns |     683.741 ns |     661.501 ns |     716.541 ns |  1.00 |    0.03 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |              5 |     683.379 ns |    17.7339 ns |  20.4224 ns |     680.776 ns |     658.646 ns |     721.854 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |              5 |      79.125 ns |     0.8675 ns |   0.7244 ns |      79.091 ns |      77.964 ns |      80.576 ns |  1.01 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |              5 |      75.768 ns |     0.6574 ns |   0.6149 ns |      75.509 ns |      75.203 ns |      77.141 ns |  0.97 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |              5 |      78.171 ns |     0.4711 ns |   0.4176 ns |      77.970 ns |      77.517 ns |      78.958 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |              1 |   1,888.142 ns |     9.6632 ns |   9.0390 ns |   1,888.478 ns |   1,864.455 ns |   1,899.102 ns |  1.02 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |              1 |   1,888.906 ns |     2.3653 ns |   2.0968 ns |   1,889.032 ns |   1,885.768 ns |   1,893.171 ns |  1.02 |    0.00 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |              1 |   1,848.599 ns |     8.6135 ns |   6.7249 ns |   1,848.834 ns |   1,838.776 ns |   1,862.185 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |              1 |     922.970 ns |     4.0907 ns |   3.6263 ns |     923.758 ns |     914.007 ns |     928.345 ns |  1.01 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |              1 |     910.176 ns |     3.6447 ns |   3.2309 ns |     910.535 ns |     902.970 ns |     914.430 ns |  0.99 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |              1 |     915.287 ns |     7.4450 ns |   6.9641 ns |     918.535 ns |     902.344 ns |     921.607 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |              5 |   6,236.907 ns |    85.6828 ns |  84.1519 ns |   6,216.353 ns |   6,152.255 ns |   6,434.916 ns |  0.96 |    0.04 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |              5 |   6,157.048 ns |   197.6720 ns | 227.6393 ns |   6,198.833 ns |   5,861.588 ns |   6,439.685 ns |  0.95 |    0.05 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |              5 |   6,466.401 ns |   233.9391 ns | 269.4046 ns |   6,389.639 ns |   6,166.614 ns |   6,979.658 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |              5 |     818.397 ns |     7.0378 ns |   6.2388 ns |     816.141 ns |     812.485 ns |     835.240 ns |  1.00 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |              5 |     835.838 ns |     5.1059 ns |   4.7761 ns |     834.511 ns |     827.863 ns |     842.235 ns |  1.02 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |              5 |     822.440 ns |     3.6806 ns |   3.4428 ns |     821.081 ns |     818.699 ns |     829.629 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 |              1 |  41,705.647 ns |   443.1338 ns | 414.5076 ns |  41,859.277 ns |  40,942.457 ns |  42,280.710 ns |  1.01 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 |              1 |  41,571.032 ns |   117.4416 ns | 104.1089 ns |  41,549.909 ns |  41,396.543 ns |  41,757.197 ns |  1.01 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 |              1 |  41,138.263 ns |   294.0126 ns | 275.0196 ns |  41,111.880 ns |  40,732.859 ns |  41,824.481 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 |              1 |  24,702.960 ns |   174.4440 ns | 145.6687 ns |  24,713.055 ns |  24,288.897 ns |  24,947.870 ns |  1.04 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 |              1 |  24,107.387 ns |   187.7083 ns | 166.3985 ns |  24,045.247 ns |  23,924.699 ns |  24,476.750 ns |  1.01 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 |              1 |  23,847.753 ns |   106.2707 ns |  99.4057 ns |  23,819.851 ns |  23,714.701 ns |  24,082.743 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 |              5 | 104,381.364 ns | 1,014.7496 ns | 949.1974 ns | 104,750.609 ns | 101,929.261 ns | 105,531.859 ns |  0.99 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 |              5 | 104,233.070 ns |   776.9727 ns | 726.7807 ns | 104,049.959 ns | 103,014.515 ns | 105,405.016 ns |  0.99 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 |              5 | 105,614.144 ns | 1,045.1556 ns | 977.6392 ns | 105,304.792 ns | 104,428.375 ns | 107,578.042 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |                |                |               |             |                |                |                |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 |              5 |  19,683.306 ns |   224.2663 ns | 209.7789 ns |  19,697.416 ns |  19,341.746 ns |  20,163.748 ns |  1.00 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 |              5 |  19,602.292 ns |    41.9998 ns |  39.2867 ns |  19,609.332 ns |  19,536.465 ns |  19,658.571 ns |  1.00 |    0.00 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 |              5 |  19,680.071 ns |   101.4223 ns |  89.9082 ns |  19,660.632 ns |  19,504.295 ns |  19,832.637 ns |  1.00 |    0.00 |         - |          NA |
```

## Perf_SingleCharFrozenDictionary
```
// * Summary *

BenchmarkDotNet v0.13.7-nightly.20230717.35, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 8.0.100-rc.2.23502.2
  [Host]     : .NET 8.0.0 (8.0.23.47906), X64 RyuJIT AVX2
  Job-RLAWOF : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-MDFFXR : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-APYSFT : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:EnableUnsafeBinaryFormatterSerialization=true  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1  

|                             Method |        Job |                         ...  Toolchain | Count |         Mean |      Error |     StdDev |       Median |          Min |          Max | Ratio | RatioSD | Allocated | Alloc Ratio |
|----------------------------------- |----------- |-------------------------...----------- |------ |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|------:|--------:|----------:|------------:|
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |     24.31 ns |   0.159 ns |   0.149 ns |     24.28 ns |     24.11 ns |     24.53 ns |  1.00 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |     37.29 ns |   0.403 ns |   0.377 ns |     37.27 ns |     36.78 ns |     38.11 ns |  1.53 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |     24.32 ns |   0.215 ns |   0.201 ns |     24.29 ns |     24.03 ns |     24.65 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |     19.55 ns |   0.122 ns |   0.115 ns |     19.59 ns |     19.30 ns |     19.71 ns |  1.00 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |     19.48 ns |   0.079 ns |   0.066 ns |     19.46 ns |     19.38 ns |     19.57 ns |  0.99 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |     19.56 ns |   0.143 ns |   0.134 ns |     19.50 ns |     19.41 ns |     19.76 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |    229.94 ns |   1.809 ns |   1.692 ns |    230.03 ns |    226.90 ns |    233.08 ns |  0.99 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |    373.39 ns |   2.665 ns |   2.226 ns |    373.54 ns |    370.58 ns |    377.73 ns |  1.61 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |    232.50 ns |   1.061 ns |   0.992 ns |    232.50 ns |    230.92 ns |    234.73 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |    186.41 ns |   1.681 ns |   1.404 ns |    185.93 ns |    185.40 ns |    190.18 ns |  1.00 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |    193.41 ns |   1.203 ns |   1.126 ns |    193.90 ns |    191.97 ns |    195.10 ns |  1.04 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |    185.77 ns |   0.541 ns |   0.480 ns |    185.68 ns |    185.22 ns |    186.93 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |  2,306.93 ns |  20.294 ns |  18.983 ns |  2,299.41 ns |  2,283.00 ns |  2,336.79 ns |  0.98 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |  3,701.93 ns |  24.969 ns |  23.356 ns |  3,690.23 ns |  3,681.71 ns |  3,747.39 ns |  1.57 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |  2,357.76 ns |  35.725 ns |  33.418 ns |  2,371.64 ns |  2,285.31 ns |  2,381.04 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |  1,884.21 ns |   4.369 ns |   3.873 ns |  1,883.50 ns |  1,878.56 ns |  1,891.05 ns |  1.00 |    0.00 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |  1,869.48 ns |  12.614 ns |  11.182 ns |  1,873.46 ns |  1,846.79 ns |  1,878.38 ns |  0.99 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |  1,888.71 ns |   3.367 ns |   2.985 ns |  1,888.59 ns |  1,884.23 ns |  1,894.37 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 | 23,377.25 ns | 300.499 ns | 281.087 ns | 23,544.00 ns | 23,005.51 ns | 23,665.87 ns |  0.99 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 | 37,639.39 ns | 261.261 ns | 218.165 ns | 37,677.77 ns | 36,941.70 ns | 37,798.24 ns |  1.60 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 | 23,523.50 ns | 267.771 ns | 250.473 ns | 23,634.59 ns | 23,098.88 ns | 23,892.40 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 | 19,459.73 ns | 207.931 ns | 194.499 ns | 19,377.97 ns | 19,177.45 ns | 19,735.74 ns |  1.02 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 | 19,391.12 ns | 159.741 ns | 149.422 ns | 19,434.71 ns | 19,084.21 ns | 19,559.08 ns |  1.02 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 | 19,086.84 ns |  44.968 ns |  37.551 ns | 19,091.02 ns | 19,024.42 ns | 19,154.67 ns |  1.00 |    0.00 |         - |          NA |
```

## Perf_SubstringFrozenDictionary
```
// * Summary *

BenchmarkDotNet v0.13.7-nightly.20230717.35, Windows 11 (10.0.22621.2428/22H2/2022Update/SunValley2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 8.0.100-rc.2.23502.2
  [Host]     : .NET 8.0.0 (8.0.23.47906), X64 RyuJIT AVX2
  Job-RLAWOF : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-MDFFXR : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-APYSFT : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:EnableUnsafeBinaryFormatterSerialization=true  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1  

|                             Method |        Job |                         ...  Toolchain | Count |         Mean |      Error |     StdDev |       Median |          Min |          Max | Ratio | RatioSD | Allocated | Alloc Ratio |
|----------------------------------- |----------- |-------------------------...----------- |------ |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|------:|--------:|----------:|------------:|
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |     38.42 ns |   0.388 ns |   0.363 ns |     38.23 ns |     37.98 ns |     39.04 ns |  0.97 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |     52.47 ns |   0.499 ns |   0.467 ns |     52.19 ns |     52.09 ns |     53.37 ns |  1.33 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |     39.40 ns |   0.441 ns |   0.391 ns |     39.19 ns |     39.08 ns |     40.36 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |    10 |     34.31 ns |   0.341 ns |   0.319 ns |     34.14 ns |     34.00 ns |     34.91 ns |  0.99 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |    10 |     34.29 ns |   0.360 ns |   0.336 ns |     34.13 ns |     33.90 ns |     34.84 ns |  0.98 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |    10 |     34.82 ns |   0.389 ns |   0.364 ns |     34.91 ns |     34.19 ns |     35.38 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |    384.91 ns |   5.106 ns |   4.776 ns |    385.28 ns |    374.92 ns |    392.21 ns |  0.99 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |    537.60 ns |   5.284 ns |   4.942 ns |    538.67 ns |    528.74 ns |    544.63 ns |  1.39 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |    386.86 ns |   1.601 ns |   1.498 ns |    387.13 ns |    382.27 ns |    388.36 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |   100 |    339.44 ns |   4.568 ns |   4.273 ns |    341.12 ns |    333.16 ns |    346.57 ns |  0.99 |    0.02 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |   100 |    343.30 ns |   5.356 ns |   5.010 ns |    346.30 ns |    333.58 ns |    348.61 ns |  1.00 |    0.02 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |   100 |    341.94 ns |   4.945 ns |   4.626 ns |    343.95 ns |    334.67 ns |    347.77 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |  4,159.28 ns |  41.013 ns |  38.363 ns |  4,173.70 ns |  4,087.19 ns |  4,218.52 ns |  1.01 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |  5,666.75 ns |  44.968 ns |  39.863 ns |  5,680.78 ns |  5,567.54 ns |  5,701.01 ns |  1.37 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |  4,126.78 ns |  33.129 ns |  30.989 ns |  4,137.46 ns |  4,069.57 ns |  4,177.16 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe |  1000 |  3,702.19 ns |  48.674 ns |  45.530 ns |  3,729.01 ns |  3,636.22 ns |  3,750.02 ns |  1.00 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe |  1000 |  3,776.44 ns |   8.614 ns |   8.058 ns |  3,775.22 ns |  3,764.91 ns |  3,793.57 ns |  1.03 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe |  1000 |  3,683.81 ns |  20.483 ns |  18.158 ns |  3,689.58 ns |  3,650.10 ns |  3,703.70 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
|  TryGetValue_True_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 | 60,042.54 ns | 677.570 ns | 633.799 ns | 59,858.66 ns | 58,924.25 ns | 60,917.46 ns |  1.01 |    0.02 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 | 78,160.59 ns | 283.472 ns | 251.290 ns | 78,039.50 ns | 77,871.91 ns | 78,678.09 ns |  1.32 |    0.01 |         - |          NA |
|  TryGetValue_True_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 | 59,371.76 ns | 699.152 ns | 653.988 ns | 59,110.17 ns | 58,657.87 ns | 60,492.30 ns |  1.00 |    0.00 |         - |          NA |
|                                    |            |                         ...            |       |              |            |            |              |              |              |       |         |           |             |
| TryGetValue_False_FrozenDictionary | Job-RLAWOF | \dotnet-runtime-classes\...corerun.exe | 10000 | 47,304.10 ns | 623.623 ns | 583.337 ns | 47,516.23 ns | 46,132.24 ns | 48,112.96 ns |  0.98 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-MDFFXR |    \dotnet-runtime-comp\...corerun.exe | 10000 | 50,309.83 ns | 798.650 ns | 747.058 ns | 50,571.26 ns | 49,405.64 ns | 51,197.12 ns |  1.04 |    0.01 |         - |          NA |
| TryGetValue_False_FrozenDictionary | Job-APYSFT |    \dotnet-runtime-main\...corerun.exe | 10000 | 48,170.73 ns | 165.619 ns | 154.920 ns | 48,129.68 ns | 47,907.78 ns | 48,481.35 ns |  1.00 |    0.00 |         - |          NA |
```

Reply:
Thanks for investigating, I guess we can keep this. Obviously more types means larger static footprints for trimmed applications but I guess `FrozenDictionary` has already jumped the shark in that regard.

cc @stephentoub @eerhardt

Reply:
I have no experience in reducing static footprints for AOT but maybe it's worth considering using the generic struct trick to force a constrained generic implementation.. does that work for AOT? IDK
 (e.g. https://github.com/dotnet/runtime/blob/e138ff137f4d727448ed171c6015905fb2d4781a/src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenSetInternalBase.cs#L272)